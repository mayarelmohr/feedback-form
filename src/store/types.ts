import { Dispatch } from "react";

// Generic types
export type FeedbackItem = {
  name: string;
  email: string;
  comment: string;
  rating: number;
};

export type FeedbackList = FeedbackItem[];

// Reducer types
export enum FeedbackActions {
  ADD = "ADD_FEEDBACK",
}

export type Action = {
  type: FeedbackActions;
  payload: FeedbackItem;
};

export type InitialStateType = {
  feedbackList: FeedbackList;
  ratingList: Array<number>;
};

// Context types
export type FeedbackContextType = {
  state: InitialStateType;
  dispatch: Dispatch<Action>;
};
