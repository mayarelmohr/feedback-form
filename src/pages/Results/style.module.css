@value colors: "styles/colors.module.css";
@value border from colors;

.layout {
  composes: layout from "styles/layout.module.css";
}

.title {
  font-weight: bold;
  font-size: 1.25em;
  margin-bottom: 10px;
}

.header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 20px;
}

.button {
  composes: button from "styles/button.module.css";
}

.list {
  list-style: none;
  margin: 0;
  padding: 0;
  max-height: 60vh;
}

.list-item:not(:last-child) {
  border-bottom: solid 1px border;
}

.review {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.review-heading {
  font-weight: bold;
  color: rgb(62, 62, 62);
  margin-bottom: 0;
}

.review-subheading {
  font-size: 0.8em;
  margin-top: 3px;
}

.chart {
  max-width: 80%;
}

@media (max-width: 768px) {
  .chart {
    max-width: 100%;
  }
}
