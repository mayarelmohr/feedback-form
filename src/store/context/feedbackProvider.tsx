import { createContext, useContext, useReducer, ReactNode, FC } from "react";
import { FeedbackContextType, InitialStateType } from "../types";
import { feedbackListReducer } from "store/reducer/feedbackReducer";
import { MAX_RATING } from "utils/constants";

// Initial state here is empty feedback list
// Rating list that has count of ratings mapping to index
// For example if we have 2 ratings for 5 stars then the array should have the number 2 at Index 4
// [0,0,0,0,2]

const initialFeedbackState: InitialStateType = {
  feedbackList: [],
  ratingList: new Array(MAX_RATING).fill(0),
};

const FeedbackListContext = createContext<FeedbackContextType>({
  state: initialFeedbackState,
  dispatch: () => null,
});

export const FeedbackListProvider: FC<{
  children: ReactNode;
  initialState?: InitialStateType;
}> = ({ children, initialState = initialFeedbackState }) => {
  const [state, dispatch] = useReducer(feedbackListReducer, initialState);
  return (
    <FeedbackListContext.Provider value={{ state, dispatch }}>
      {children}
    </FeedbackListContext.Provider>
  );
};

export const useFeedbackListContext = () => {
  return useContext(FeedbackListContext);
};
