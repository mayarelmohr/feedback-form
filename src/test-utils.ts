import { screen } from "@testing-library/react";

export const getTextField = (name: string) => {
  return screen.getByRole("textbox", { name: new RegExp(name, "i") });
};
