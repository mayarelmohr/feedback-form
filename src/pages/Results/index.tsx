import { useNavigate } from "react-router-dom";
import classNames from "classnames/bind";
import styles from "./style.module.css";
import { useFeedbackListContext } from "store/context/feedbackProvider";
import Chart from "components/Chart";
import Rating from "components/Rating";

const cx = classNames.bind(styles);

const Results = () => {
  const navigate = useNavigate();
  const {
    state: { feedbackList },
  } = useFeedbackListContext();

  return (
    <div className={cx("layout")}>
      <div className={cx("header")}>
        <h2 className={cx("title")}>Feedback results</h2>
        <button onClick={() => navigate(-1)} className={cx("button")}>
          Go back
        </button>
      </div>
      <div className={cx("chart")}>
        <Chart />
      </div>
      <h3>User Reviews</h3>
      {feedbackList.length < 1 ? (
        <p>There are no results</p>
      ) : (
        <ul className={cx("list")}>
          {feedbackList.map((item) => (
            <li className={cx("list-item")} key={item.email}>
              <div className={cx("review")}>
                <div>
                  <p className={cx("review-heading")}>{item.name}</p>
                  <p className={cx("review-subheading")}>{item.email}</p>
                </div>
                <Rating rate={item.rating} />
              </div>
              <p>{item.comment}</p>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default Results;
