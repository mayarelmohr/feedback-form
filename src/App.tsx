import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import "./App.css";
import FeedbackForm from "pages/Home";
import Results from "pages/Results";
import Header from "components/Header";
import { FeedbackListProvider } from "store/context/feedbackProvider";

function App() {
  return (
    <Router>
      <Header />
      <main>
        <FeedbackListProvider>
          <Routes>
            <Route path="/" element={<FeedbackForm />} />
            <Route path="/results" element={<Results />} />
          </Routes>
        </FeedbackListProvider>
      </main>
    </Router>
  );
}

export default App;
