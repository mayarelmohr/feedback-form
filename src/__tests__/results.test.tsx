import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Results from "pages/Results";
import { getTextField } from "test-utils";
import userEvent from "@testing-library/user-event";
import { FeedbackListProvider } from "store/context/feedbackProvider";
import { InitialStateType } from "store/types";
import App from "App";

const initialState: InitialStateType = {
  feedbackList: [
    {
      name: "john doe",
      email: "john.doe@gmail.com",
      rating: 4,
      comment:
        "Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor quod impedit alias vitae molestias aperiam animi sunt tenetur harum molestiae repellat dolorum, exercitationem veritatis reprehenderit obcaecati. Ratione quaerat minus odit.",
    },
  ],
  ratingList: [0, 0, 0, 1, 0],
};

const initialEmptyState: InitialStateType = {
  feedbackList: [],
  ratingList: [0, 0, 0, 0, 0],
};

const renderResultsPageWithContext = (initialState: InitialStateType) => {
  return render(
    <FeedbackListProvider initialState={initialState}>
      <BrowserRouter>
        <Results />
      </BrowserRouter>
    </FeedbackListProvider>
  );
};

describe("Feedback results", () => {
  it("Results page matches snapshot", () => {
    const { container } = renderResultsPageWithContext(initialState);
    expect(container).toMatchSnapshot();
  });

  it("Go back button shows an empty form", async () => {
    render(<App />);
    userEvent.type(getTextField("name"), "John");
    userEvent.type(getTextField("email"), "john.doe@gmail.com");
    userEvent.type(getTextField("comment"), "Amazing product");
    userEvent.click(screen.getByTestId("2-stars"));
    userEvent.click(screen.getByText(/submit feedback/i));
    userEvent.click(screen.getByText(/go back/i));
    const heading = await screen.findByRole("heading", {
      name: /add your feedback/i,
    });
    expect(heading).toBeInTheDocument();
    expect(getTextField("name")).toHaveValue("");
    expect(getTextField("email")).toHaveValue("");
    expect(getTextField("comment")).toHaveValue("");
    const ratings = screen.getAllByRole("radio");
    ratings.forEach((rating) => {
      expect(rating).not.toBeChecked();
    });
  });

  it("shows list of feedbacks", () => {
    renderResultsPageWithContext(initialState);
    expect(
      screen.getByText(initialState.feedbackList[0].name)
    ).toBeInTheDocument();
    expect(
      screen.getByText(initialState.feedbackList[0].email)
    ).toBeInTheDocument();
    expect(
      screen.getByText(initialState.feedbackList[0].comment)
    ).toBeInTheDocument();
    expect(
      screen.getByTestId(`rating-${initialState.feedbackList[0].rating}-stars`)
    ).toBeInTheDocument();
  });

  it("shows no results when feedback is empty", () => {
    renderResultsPageWithContext(initialEmptyState);
    expect(screen.getByText(/there are no results/i)).toBeInTheDocument();
  });
});
