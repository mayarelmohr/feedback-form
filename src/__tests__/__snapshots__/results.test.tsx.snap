// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback results Results page matches snapshot 1`] = `
<div>
  <div
    class="layout"
  >
    <div
      class="header"
    >
      <h2
        class="title"
      >
        Feedback results
      </h2>
      <button
        class="button"
      >
        Go back
      </button>
    </div>
    <div
      class="chart"
    >
      Chart
    </div>
    <h3>
      User Reviews
    </h3>
    <ul
      class="list"
    >
      <li
        class="list-item"
      >
        <div
          class="review"
        >
          <div>
            <p
              class="review-heading"
            >
              john doe
            </p>
            <p
              class="review-subheading"
            >
              john.doe@gmail.com
            </p>
          </div>
          <div
            aria-label="Rating: 4 out of 5 stars"
            data-testid="rating-4-stars"
            role="img"
          >
            <span
              aria-hidden="true"
            >
              <svg
                class="rating rating-wrapper filled"
              >
                star.svg
              </svg>
            </span>
            <span
              aria-hidden="true"
            >
              <svg
                class="rating rating-wrapper filled"
              >
                star.svg
              </svg>
            </span>
            <span
              aria-hidden="true"
            >
              <svg
                class="rating rating-wrapper filled"
              >
                star.svg
              </svg>
            </span>
            <span
              aria-hidden="true"
            >
              <svg
                class="rating rating-wrapper filled"
              >
                star.svg
              </svg>
            </span>
            <span
              aria-hidden="true"
            >
              <svg
                class="rating rating-wrapper"
              >
                star.svg
              </svg>
            </span>
          </div>
        </div>
        <p>
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor quod impedit alias vitae molestias aperiam animi sunt tenetur harum molestiae repellat dolorum, exercitationem veritatis reprehenderit obcaecati. Ratione quaerat minus odit.
        </p>
      </li>
    </ul>
  </div>
</div>
`;
